### PR TITLE
Check for overflow on left shift of signed ints

### PR DESCRIPTION
**Pull request by tautschnig**
This is undefined behaviour according to 6.5.7 paragraph 4 in C11.

### CODE REVIEW

**Comment by kroening** [comment]
BTW, clang also warns:
warning: shifting a negative signed value is undefined
What does the standard say?

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 1]
```diff
@@ -96,6 +95,43 @@ literalt boolbvt::convert_overflow(const exprt &expr)
...
+    bvt result=bv_utils.shift(bv_ext, bv_utilst::shiftt::LEFT, bv1);
+
+    // an undefined shift of a non-zero value always results in overflow
+    literalt undef=
+      bv_utils.rel(
+        bv1,
+        ID_gt,
+        bv_utils.build_constant(old_size, bv1.size()),
+        bv_utilst::representationt::UNSIGNED);
+
```
Comment (kroening) [inline src/solvers/flattening/boolbv_overflow.cpp:132]: I believe that the right-hand-side operand may be either signed or unsigned. I think the above will only work if the right-hand side has enough bits (i.e., sign extension may be needed).

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 2]
```diff
@@ -96,6 +95,43 @@ literalt boolbvt::convert_overflow(const exprt &expr)
...
+    {
+      // get top result bits
+      result.erase(result.begin(), result.begin()+old_size);
+    }
+    else
+    {
+      // get top result bits plus sign bit
+      DATA_INVARIANT(old_size!=0, "zero-size operand");
+      result.erase(result.begin(), result.begin()+old_size-1);
+    }
```
Comment (kroening) [inline src/solvers/flattening/boolbv_overflow.cpp]: Did you intend that left-shifts of negative numbers are ruled out (see other comment)?

File: src/solvers/flattening/boolbv_overflow.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_overflow.cpp:132]: I was thinking that negative shifts should not be caught here, but indeed I need to make sure that this case is truly ignored.

Comment (tautschnig) [reply src/solvers/flattening/boolbv_overflow.cpp:132]: I've added this check in the updated revision.

**Comment by tautschnig** [comment]
> BTW, clang also warns:
warning: shifting a negative signed value is undefined
What does the standard say?

Where is it that this warning appears?

**Comment by kroening** [comment]
Say try
printf("%d\n", -1<<((sizeof(int)*8-1)));

File: regression/cbmc/Overflow_Subtraction1/main.c [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#define INT_MIN (-1<<(sizeof(int)*8-1))
+#define INT_MIN (1U<<(sizeof(int)*8-1))
```
Comment (kroening) [inline regression/cbmc/Overflow_Subtraction1/main.c]: This is now at least an unusual definition of INT_MIN

**Comment by kroening** [comment]
The following seems to be the right approach on shifting negative numbers to the left (do not confuse with negative shift distances): It's an overflow iff a) the new sign is 0, or b) any of the bits that are shifted out to the left are 0.

File: regression/cbmc/Overflow_Subtraction1/main.c (hunk 3 above)
Comment (tautschnig) [reply regression/cbmc/Overflow_Subtraction1/main.c]: Yet it has defined behaviour. Some discussion can also be found at https://blog.regehr.org/archives/738. Actual header files appear to be using `-INT_MAX-1`.

Comment (tautschnig) [reply regression/cbmc/Overflow_Subtraction1/main.c]: Reverted/fixed in the latest revision.

**Comment by tautschnig** [comment]
> The following seems to be the right approach on shifting negative numbers to the left (do not confuse with negative shift distances): It's an overflow iff a) the new sign is 0, or b) any of the bits that are shifted out to the left are 0.

Thank you, you are of course right. This has now been addressed.

**Comment by tautschnig** [comment]
> printf("%d\n", -1<<((sizeof(int)*8-1)));

It seems actually all shifts of negative numbers are undefined:
"If E1 has a signed type and nonnegative value, and E1 × 2^E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined."

Maybe the question is: should negative numbers be flagged as overflows or as invalid shifts?

**Comment by peterschrammel** [comment]
I think shifts of negative numbers should be invalid shifts.

**Comment by kroening** [comment]
I would also go for "invalid shift"

**Comment by tautschnig** [comment]
@peterschrammel @kroening Done in 1e3712c.

**Pull request merged**